### DISCUSSION TITLE
Reconsider nodes management

### DISCUSSION THREAD

**Discussion by ralphwetzel** [phoddie/node-red-mcu#16] [General]
As of today, there are a number of `node` implementations that have to play together, to make `node-red-mcu` work. The following table tries to display what I am aware of, serving as a starting point to discuss how to organize this for the future into a sustainable concept.

\# | implementation | required by | located @ | purpose
--- | --- | --- | --- | ---
1 | `sample-node.html` | Node-RED | `.node-red/node_modules/...` | UI @ the node-RED editor
2 | `sample-node.js` | Node-RED | `.node-red/node_modules/...` | implementation of the node's functionality for Node-RED runtime
3 | `class SampleNode` | node-red-mcu | `node-red-mcu/nodered.js` | implementation of the node's functionality for node-red-mcu
4 | `mcu*sample-node` | node-red-mcu-plugin | `node-red-mcu-plugin/mcu-plugin.js` | replacing `sample-node.js` in the Node-RED runtime to interface w/ node-red-mcu

[To be fair: This is a simplification -  as there could be several node implementations within one `sample-node.js` file.]

Currently, there's an established concept for 1 & 2.
I'm proposing to discuss how to organize 3 & 4 (perhaps in combination w/ 1 and 2) and as well e.g. if to publish and install them as npm packages.
Consequentially this has influence on other aspects as well, e.g. the build environment setup.
Any ideas on how to proceed?

**Comment by phoddie** [comment]
Thanks for starting this conversation.

There are core (built-in) nodes and contributed (plug-in or external) nodes. This thread is about the contributed nodes only.

You list four components of a contributed node. There a few more:

5. Code to be used by `nodered2mcu` to optimize the configuration for the MCU. There's no standalone example of this as `nodered2mcu` doesn't support this capability yet.
6. The `manifest.json` used to build the MCU code
7. There's also the `package.json` used from npm

Let's use the classic `lower-case` example for discussion. Its directory contains:

- `lower-case.html` (1)
- `lower-case.js` (2)
- `package.json` (7)

I propose an `mcu` subdirectory to contain all files solely used with Node-RED MCU Edition. The presence of the `mcu` directory is an indication that this contributed node is compatible with the Node-RED MCU Edition. The `mcu` directory contains:

- `manifest.json` (6) to build the node. This includes any dependencies (for example, the MCU [websocket node manifest](https://github.com/phoddie/node-red-mcu/blob/main/nodes/websocketnodes/manifest.json) includes the ECMA-419 WebSocket manifest)
- The source files and assets specifically for the MCU runtime (3). Note that there may be none if the module implementation for the full Node-RED (2) may also be used by the MCU.
- An optional JavaScript source file to transform the configuration in `nodered2mcu` (5). It may be enough for this to have a well known file name such as `node2mcu.js`.

As part of assembling the project's `manifest.json`, the tooling (your Node-RED MCU Builder plug-in) will update the `includes` section by adding the manifest for each plug-in node. This will optimize the code footprint of the device binary by including only those plug-in nodes used by the project.

**Reply by ralphwetzel** [reply]
That's a good starting point. Thank you!

> As part of assembling the project's `manifest.json`, the tooling (your Node-RED MCU Builder plug-in) will update the `includes` section by adding the manifest for each plug-in node.

Let's see what is necessary to achieve this...

**Reply by ralphwetzel** [reply]
A question @phoddie related to the `manifest` definition:

The [documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/manifest.md#strip) states the `strip` key.
The `manifest.json` setup for `node-red-mcu` uses `x-strip`:

https://github.com/phoddie/node-red-mcu/blob/6072a35daddf1d36b8f910c9266d9f8512c4ad3d/manifest.json#L35-L39

What's the difference between those - and is `x-strip` a valid key of `manifest.json`?

**Reply by phoddie** [reply]
`x-strip` does nothing at all. It just turns that block into a comment so that it is ignored by `mcconfig`. That block is left over from some experiments during development. It could just be removed for now, though we may eventually want it to support Node-RED [Mods](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/mods.md).

**Reply by ralphwetzel** [reply]
@phoddie 
As I'm auto-building the `manifest.json`, I come across the situation, that developers chose an extension different to `.js` to name the source file of their modules (e.g. `.es.js`). Is there a way to tell the compiler to accept those files as a source file of module `module`, e.g. a translation table in the sense of "take this file as source for a module named `module`"?
```
  "modules": {
    "*": [ "/some/path/module.es.js" ]
  }
```
... creates a compiler warning (as it declares the file & not the module).

**Reply by ralphwetzel** [reply]
There are some additional questions that came up:

1) Staying with the example of
```
  "modules": {
    "*": [ "/some/path/module.es.js" ]
  }
```
one could be tempted to declare the module as `module.es`.
```
  "modules": {
    "*": [ "/some/path/module.es" ]
  }
```
This compiles, despite the module is referenced in the importing / requiring (parent-)module as `module` - thus breaking the code. 👎 Any proposals?

2) I have a case, where a `.map` file is present next to the module's file:
```
module.js
module.js.map
```
Although it's not utilized by `module.js` at all, the compiler tries to process it ... and claims e.g. missing files that are referenced in `module.js.map`. Is there a way to prevent this?

**Reply by phoddie** [reply]
@ralphwetzel – the manifest does not include the file extension. `mcconfig` finds files that match (could be `.js`, `.ts`, `.c`, etc.). Your example above builds if the `.js` is excluded:

```json
  "modules": {
    "*": [ "./example.es" ]
  }
```

The default module specifier is derived from the filename with the file extension removed. Hence, here the module is importable using the module specifier "example.es":

```js
import example from "example.es"
```

To use the module specifier "example" instead, remap in the manifest:

```json
  "modules": {
    "example": "./example.es"
  }
```

```js
import example from "example"
```

Regarding the issue you report with the `*.map` file, I'm unsure how to reproduce that. I'd be happy to take if a look if you would share a test case. Thank you.

**Reply by ralphwetzel** [reply]
@phoddie: I'm changing the way to create the build directory. Whereas I used to copy "everything" from node-red-mcu into that (temporary) directory, I intend to "just" generate a `manifest.json` and use this to reference to the one [defined by `node-red-mcu`](https://github.com/phoddie/node-red-mcu/blob/main/manifest.json). That works perfect - despite one aspect:

The given `manifest.json` references a [`flows.json`](https://github.com/phoddie/node-red-mcu/blob/main/flows.json) in the same directory:

https://github.com/phoddie/node-red-mcu/blob/01470d80b77d402be876fdc2e80ea36752234d68/manifest.json#L19-L28

I yet intend to operate with different `flows.json` file (located in the created build directory). As consequence, I end up with two `flow.json`s that are processed by `nodered2mcu` ... and the first (empty) file wins (currently).

I tried to opt-out the `flows.json` of the `node-red-mcu` directory using the "~" option ... without success:

```
"modules": {
    "*": [
      {
        "source": "./flows",
        "transform": "nodered2mcu"
      }
    ],
    "~": "$(MCUROOT)/flows"
  }
```

Did I fail to understand the idea behind `"~"` ... or do you spot another error?

**Reply by phoddie** [reply]
The `~` feature of manifests is always a little tricky. Here I don't think we should use it. Rather, we should have a manifest specifically for the Node-RED MCU runtime and a separate application manifest that includes it. That makes your work simpler. I committed a first pass at that. the `manifest_runtime.json` file does not include `main` or `flows` so you can organize those however you like.

**Reply by ralphwetzel** [reply]
> The `~` feature of manifests is always a little tricky

It is - and I made it (finally) working. In this case, the reference to the `transform` property is necessary:

```
"~": [
      {
        "source": "$(MCUROOT)/flows",
        "transform": "nodered2mcu"
      }
```

Anyway. The setup becomes fairly cleaner when splitting the manifests. Thank you for that step!

**Reply by phoddie** [reply]
Glad that worked.

**Comment by ralphwetzel** [comment]
@phoddie:
I'm still working through the jungle of auto-building manifests 😄, currently sorting out some details:

When including a module that has a `.ts` file next to a `.js` in the same directory, `mcconfig` is generating two (2) make commands, the second one to compile the `.ts` file to `.js`:

1)
``` makefile
$(MODULES_DIR)/[...]/logger.xsb: [...]/logger/src/index.js
	@echo "# xsc [...]/logger.xsb"
	$(XSC) [...]/logger/src/index.js -d -c -e -o $(@D) -r logger
```

2)
``` makefile
$(MODULES_DIR)/[...]/logger.xsb: $(MODULES_DIR)/index.js
	@echo "# xsc [...]/logger.xsb"
	$(XSC) $(MODULES_DIR)/index.js -d -c -e -o $(@D) -r logger
%/index.js : %/tsconfig.json 
	@echo "# tsc tsconfig.json"
	tsc -p $(MODULES_DIR)/tsconfig.json
```


This triggers two warnings
```
[...]/makefile:1019: warning: overriding commands for target `[...]/logger.xsb'
[...]/makefile:787: warning: ignoring old commands for target `[...]/logger.xsb'
```
and then fails w/ `make: tsc: No such file or directory` ... as there's no path to `tsc` given.

In this case, the `tsc` action is not necessary - as the `.js` file is already in place.

Is there a flag to tell `mcconfig` **not to care for** `.ts` files ... preferably definable per `manifest.json`?

**Reply by phoddie** [reply]
@ralphwetzel – Thanks for the report from the jungle. You have bumped into an interesting problem. As you know, the manifest doesn't contain file extensions and `mcconfig` figures out what to do based on the file matches it finds. This is often convenient -- a module with both JavaScript and C implementations is included as "./module" and both "module.c" and "module.js" are built. We didn't anticipate a situation where there would be two different implementations of the same code (e.g. "./module.ts" and "module.js"). In this case, I'm assuming you are working with a package that provides both the TypeScript source and the transpiled JavaScript version of that. Some options when both are present:

- Always use the JavaScript version. This will work even if TypeScript is not installed.
- Always use the TypeScript version. This makes the type information available to other TypeScript sources in the project.
- Fail, since the situation is ambiguous. This is very safe but won't work with some packages.
- Provide a way in the manifest to specify which to use -- this could be required when there is a conflict or could be optional with some default (e.g. use the TypeScript version)

I'm not sure.... what do you think?

**Reply by ralphwetzel** [reply]
Indeed, your current strategy is very convenient & powerful seen from the developer's perspective. As this is a very edge case, I propose to implement a way to opt-out the transpilation of the TypeScript source - when & after the above mentioned error was hit. I've seen there's already a `typescript` property [defined for `manifest.json`](https://github.com/Moddable-OpenSource/moddable/blob/ae5ed20a963b22646f2d512101dff526006ae49d/tools/mcmanifest.js#L1850-L1866). Based on this, a flag like

```
typescript: { "compiler": "false" }
```
would fully meet my (current) demands.

There could be alternative implementations as well:

- Check if the potential result of the `tsc` operation is already present (in the same directory) & then skip the transpilation.
- Provide a way to opt-out _files_ from being processed at all.

The first one fits into the idea of

> `mcconfig` figures out what to do based on the file matches it finds

whereas I'd prefer to have a bit more control over what's happening in that case.
The second one has further potential beyond this issue, yet requires some more effort for implementation.

**Comment by ralphwetzel** [comment]
I've pushed the next iteration of the plugin into the repo. There's as well a bit of documentation that discusses what the plugin does regarding the manifest creation process. Just be warned: The number of loose ends & incomplete implementations is significant.

**Reply by phoddie** [reply]
@ralphwetzel – apologies for being quiet here. I've been working through some complicated issues in the runtime. The behavior of `done()` for nodes now matches, as best I understand it, full Node-RED. That is necessary to be able to correctly run implementations of nodes from the full Node-RED (`delay`, `join`, and `trigger` for example). In addition, this work allowed the implementation of `Complete` and `Catch` nodes to be optimized considerably which is important because otherwise supporting them can be very expensive. That work is all committed.

I'd like to get back to supporting the plug-in development. I'll make some time to review your current commit and documentation. Is there something specific I can help with beyond that?

**Reply by ralphwetzel** [reply]
@phoddie - it's great to follow the constant progress of node-red-mcu. I'm testing the plugin & extending it's capabilities step by step; currently I'm yet lacking time to finalize the next iteration for release. Nevertheless your feedback is highly welcome...

**Reply by phoddie** [reply]
Thanks so much. I will do that this evening.

**Reply by phoddie** [reply]
I updated node-red-mcu-plugin in my Node-RED installation. The UX works, but building fails:

```
# cc mc.resources.c
1 warning generated.
ReferenceError: (host): module "structuredClone" not found
### ReferenceError: (host): module "structuredClone" not found!
# exception: (host): module "structuredClone" not found!
```

The version of the node-red-mcu sources in `<userDir>/mcu-plugin-cache` is out-of-date. That could explain the failure. What do you suggest?

**Reply by ralphwetzel** [reply]
@phoddie - how did you update the plugin?

```
cd <path to your .node-red folder>
npm install https://github.com/ralphwetzel/node-red-mcu-plugin
```
... should pull the latest version of `node-red-mcu` from its github repository ...
... into `node_modules/@ralphwetzel/node-red-mcu-plugin/node-red-mcu`.

If the files there are outdated, you could move them by hand ... and I would try to identify the issue.

**Reply by phoddie** [reply]
I executed the two lines you show. Maybe I misunderstood `<path to your .node-red folder>`. I used the path where `node-red` resolves to on my computer (` /usr/local/lib/node_modules/node-red`) but maybe I should have used `<userDir>` (`/Users/hoddie/.node-red`) instead? I can try that later.

**Reply by ralphwetzel** [reply]
It should be `<userDir>`. I'll clarify this in the documentation...

**Reply by phoddie** [reply]
Good news. It works for me again. The main thing was installing in the correct directory.
- I was getting connection errors from the proxy. I had to manual change the port in xsbug from 5002 to 5004. I recall we added environment variables so that manual configuration wouldn't be necessary. Is that not working?
- I never remember to Deploy in Node-RED before Building for the MCU.
- Inject isn't working. How can I help with that? It is a really valuable feature for testing. I set a breakpoint in the global receive function for those messages `<xsbug:script>` but it is never hit.

If it would help. perhaps we could select a time to do some realtime work with screen sharing over Zoom. I think you have my email address. Feel free to reach out, if that might be useful.

**Reply by ralphwetzel** [reply]
> Good news. It works for me again.

👍 

> * I was getting connection errors from the proxy. I had to manual change the port in xsbug from 5002 to 5004. I recall we added environment variables so that manual configuration wouldn't be necessary. Is that not working?

The `xsbug` connection will be initiated if you engage the debug mode in the _Configuration_ panel @ _Instrumentation_:

<img width="274" alt="image" src="https://user-images.githubusercontent.com/16342003/191574077-f726b176-29f0-4ba3-b323-536cafe87123.png">

(I'm aware that this part of the GUI needs an update; already on my ToDo list.)

This setting should be translated into a command line parameter for `mcconfig`. You may check that via the _Console Monitor_, e.g.:

`> mcconfig -d -x localhost:5004 -m -p esp32/m5atom_lite`

> * I never remember to Deploy in Node-RED before Building for the MCU.

I have an idea of how to address this...

> * Inject isn't working. How can I help with that? It is a really valuable feature for testing. I set a breakpoint in the global receive function for those messages `<xsbug:script>` but it is never hit.

This could be related to the other issue. I'll try to investigate...

I'll send you a PM - assuming that this will make communication a bit more straight!

**Reply by phoddie** [reply]
I have "Debug" checked in Instrumentation as shown. Most of the time, the proxy fails when trying to run the project:

```
@manual_build
Type "debug" = core node: No manifest added.
Type "inject" = core node: No manifest added.
mcuProxy: Listening!
21 Sep 22:20:56 - [red] Uncaught Exception:
21 Sep 22:20:56 - [error] Error: connect ECONNREFUSED ::1:5002
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1300:16)
```

xsbug is listening on 5002 (the default). The error occurs whether or not xsbug is launched before starting the build.

**Reply by ralphwetzel** [reply]
This could be an [issue](https://github.com/nodejs/node/issues/40702) of how your system translates `localhost` to an IP address. Which version of `node.js` are you working with?

I've pushed a mod to `proxy.js` that is a bit more verbose, logging the two connection parameter sets. On my system this reads:

```
    22 Sep 20:55:55 - [info] Node-RED Version: v3.0.2
    22 Sep 20:55:55 - [info] Node.js  Version: v16.13.1
    22 Sep 20:55:55 - [info] Darwin 21.6.0 arm64 LE
[...]
    Launching mcuProxy:
    - Listening for MCU @ :::5004/IPv6
    - Connecting to xsbug: Resolved localhost to 127.0.0.1/4.
```

**Reply by phoddie** [reply]
I'm running a later Node.js than you, but an earlier Node-RED (!):
```
22 Sep 12:26:45 - [info] Node-RED version: v2.2.2
22 Sep 12:26:45 - [info] Node.js  version: v18.8.0
22 Sep 12:26:45 - [info] Darwin 21.6.0 arm64 LE
```
I'll give your changes a try later. Meanwhile, should I update to Node-RED v3.x?? I didn't dare because I don't really understand what changed. But, it seems to work for you. Maybe that explains some of what is different in our results.

**Reply by ralphwetzel** [reply]
The change referenced in the mentioned issue was introduced w/ Node17 - and could make a difference.
Node-RED 3.0.2 works like a charm - yet hasn't changed significantly with respect to the core functionalities.
My proposal: Keep your system as it is & let's try to find the cause for this annoyance first. I'm pretty sure, others might face this as well at some point in time.

**Reply by phoddie** [reply]
Mixed results with the latest.

- Now when I forget to deploy before building, there is a helpful dialog that asks if I want to deploy first. Very nice. Thank you!
- I got an error about missing `loglevel` module when launching `node-red`. I worked around that by `npm i loglevel`. Maybe a missing dependency?
- The proxy isn't working yet, but the additional logging may offer some clues for you:
```
Launching mcuProxy:
- Listening for MCU @ :::5004/IPv6
- Connecting to xsbug: Resolved localhost to ::1/6.
```

**Reply by ralphwetzel** [reply]
> * The proxy isn't working yet, but the additional logging may offer some clues for you:
> ````
> - Connecting to xsbug: Resolved localhost to ::1/6.

That confirms my assumption. Should be fixed soon...

**Reply by ralphwetzel** [reply]
> Should be fixed soon...

Is fixed now.

**Reply by phoddie** [reply]
It is! So great! I used Build Target "All". With that, it deploys, xsbug works, inject works from the Node-RED Editor, and debug messages are sent to the Node-RED Editor. Thank you!

**Reply by phoddie** [reply]
There is an interesting second-order bug. After running a flow on the MCU, I can no longer run that flow in Node-RED. When I try, it deploys but no debug output appears and inject seems to be ignored. If I manually remove the `_mcu` properties from the nodes, then it works again with Node-RED.

**Reply by ralphwetzel** [reply]
> There is an interesting second-order bug. After running a flow on the MCU, I can no longer run that flow in Node-RED.

Do you still have this issue with the latest changes of the plugin?
